internal/pkg/cli/deploy/env.go

```diff
@@ -100,11 +141,115 @@ func NewEnvDeployer(in *NewEnvDeployerInput) (*envDeployer, error) {
+			return cfnstack.NewEnvConfigFromExistingStack(in, lastForceUpdateID, oldParams)
+		},
+		envDescriber:      envDescriber,
+		envManagerSession: envManagerSession,
+		newLBDescriber: func(sess *session.Session) lbDescriber {
+			return elbv2.New(sess)
+		},
+		newServiceStackDescriber: func(svc string, sess *session.Session) stackDescriber {
+			return stack.NewStackDescriber(cfnstack.NameForService(in.App.Name, in.Env.Name, svc), sess)
 		},
 	}, nil
 }
 
+// Verify checks that the manifest configuration is valid to deploy alongside
+// the currently deployed resources.
+func (d *envDeployer) Verify(ctx context.Context, mft *manifest.Environment) error {
+	if mft.CDNEnabled() && mft.HTTPConfig.Public.Certificates == nil && d.app.Domain != "" {
+		// With managed domain, if the customer isn't using `alias` the A-records are inserted in the service stack as each service domain is unique.
+		// However, when clients enable CloudFront, they would need to update all their existing records to now point to the distribution.
+		// Hence, we force users to use `alias` and let the records be written in the environment stack instead.
+		if err := d.envDescriber.ValidateCFServiceDomainAliases(); err != nil {
+			return err
+		}
+	}
+
+	if mft.CDNEnabled() && aws.BoolValue(mft.CDNConfig.Config.TerminateTLS) {
+		// ensure all services _are not_ doing http->https redirect
+		if err := d.verifyALBWorkloadsDontRedirect(ctx); err != nil {
+			return fmt.Errorf("can't enable TLS termination on CDN: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// verifyALBWorkloadsDontRedirect verifies that none of the ALB Workloads
+// in this environment have a redirect in their HTTPWithDomain listener.
+// If any services redirect, an error is returned.
+func (d *envDeployer) verifyALBWorkloadsDontRedirect(ctx context.Context) error {
+	params, err := d.envDescriber.Params()
+	if err != nil {
+		return fmt.Errorf("get env params: %w", err)
+	}
+
+	services := strings.Split(params[cfnstack.EnvParamALBWorkloadsKey], ",")
+	g, ctx := errgroup.WithContext(ctx)
+
+	var badServices []string
+	var badServicesMu sync.Mutex
+
+	for i := range services {
+		svc := services[i]
+		g.Go(func() error {
+			redirects, err := d.lbServiceRedirects(ctx, svc)
+			switch {
+			case err != nil:
+				return fmt.Errorf("verify service %q: %w", svc, err)
+			case redirects:
+				badServicesMu.Lock()
+				defer badServicesMu.Unlock()
+				badServices = append(badServices, svc)
+			}
+
+			return nil
+		})
+	}
+
+	if err := g.Wait(); err != nil {
+		return err
+	}
+	if len(badServices) > 0 {
+		sort.Strings(badServices)
+		return fmt.Errorf("HTTP traffic redirects to HTTPS in %v %v.\nSet http.redirect_to_https to false for %v and redeploy %v.",
+			english.PluralWord(len(badServices), "service", "services"),
+			english.OxfordWordSeries(badServices, "and"),
+			english.PluralWord(len(badServices), "that service", "those services"),
+			english.PluralWord(len(badServices), "it", "them"),
+		)
```

Comment:
Really nice error message <img src="https://emojis.slackmojis.com/emojis/images/1643514239/2076/in_love.gif?1643514239" width=25px />

instead of having a new line in the error mesage, what do you think of adding a custom error type that implements the following interface:
https://github.com/aws/copilot-cli/blob/6132cb54fb1167b297bce0a218026760cfbae0a0/cmd/copilot/main.go#L19-L21

```go
type errEnvHasPublicServicesWithRedirect struct  { /* ... */ }

func (e errEnvHasPublicServicesWithRedirect) Error() string {
   return fmt.Sprintf("services %q redirect HTTP traffic to HTTPS"
}

func (e errEnvHasPublicServicesWithRedirect) RecommendedAction() string {
   return fmt.Sprintf(`In the service manifests for %q, set the following field:
%s
and run %s`, naughtyServices, color.HighlightCodeBlock(`http:
  redirect_to_https: true
`), color.HighlightCode("copilot svc deploy") 
}
```